README.md
# LDJump
**LDJump** is an R package to estimate variable recombination rates from population genetic data. 
It is a unix based program (with a necessary installation of [PhiPack](<https://www.maths.otago.ac.nz/~dbryant/software.html>)), able to estimate the recombination map of sequences in fasta and vcf format. 
First, the sequences are divided in short segments of user defined length. The constant recombination rate is estimated for every segment with a regression model. 
This set of estimates is fed in a segmentation algorithm (SMUCE) to estimate the breakpoints of the recombination landscape. A [PDF Manual](./LDJump.pdf) with complete documentation of each function is also available. The publication introducing **LDJump** can be found [here](<https://onlinelibrary.wiley.com/doi/abs/10.1111/1755-0998.12994?af=R>).

### Author (Requests)
Please contact me in case of questions, comments, bug reports, etc...

    Author: Philipp Hermann
    E-Mail: philipp.hermann@jku.at

## Dependencies & System Requirements
This package makes use of several functions of other R-packages, of the software package [PhiPack](<https://www.maths.otago.ac.nz/~dbryant/software.html>), and it can optionally also make use of [LDhat](<https://github.com/auton1/LDhat>) in order to decrease the runtime of **LDJump**. Here we provide a list of used R-packages, as well as, other software packages: 

* Unix Operating System
* R (>= 2.10)
* adegenet (>= 2.0.1)
* ape
* genetics (>= 1.3.8.1)
* Biostrings (>= 2.38.4)
* stepR
* seqinr (>= 3.1-3)
* quantreg
* vcfR (>= 1.5.0)
* PhiPack
* LDhat (2.2, optional -> to decrease runtime)
* snow
* scrm (1.7.1, optional to train the regression model under different scenarios than considered)
* ms2dna (1.16, optional to train the regression model under different scenarios than considered)

Notice that **Biostrings** needs to be installed via [Bioconductor](<http://bioconductor.org/packages/release/bioc/html/Biostrings.html>).  [PhiPack](<https://www.maths.otago.ac.nz/~dbryant/software.html>) as well as the functions *dos2unix* and *awk* neeed to be installed too. Additionally, we recommend to also install [LDhat](<https://github.com/auton1/LDhat>) which will then enable to compute estimates at lower computational cost. The software packages [scrm](<https://github.com/scrm/scrm>) and [ms2dna](<http://guanine.evolbio.mpg.de/bioBox/>)) only need to be installed when the user wants to estimate a regression model based on simulations under a user input demographic scenario, see [Update to Version 0.2.2](<###Update to Version 0.2.2>).


## Installation
The ZIP-File of the package can be downloaded via "Clone or download" as well as the command line: 

```markdown
git clone https://github.com/PhHermann/LDJump.git
R CMD build LDJump
R CMD INSTALL LDJump_<version>.tar.gz
``` 

Using *R* one can also install the *.tar.gz* file with the following command: 
```R
install.packages("/PathToLDJump/LDJump_<version>.tar.gz", repos=NULL, type="source")
```

## Usage

After loading the package in the workspace one can use the main function *LDJump* in order to estimate the variable recombination rate for the population of interest. We recommend to use **LDJump** with the population in *fasta*-Format. Alternatively, restructuring of *vcf*-Files is also possible with a implemented function *vcfR_to_fasta*. Therefore, we used the reference sequence from <http://phase3browser.1000genomes.org/Homo_sapiens> for our example: 

```R
require(LDJump)
LDJump(seqName, alpha = 0.05, segLength = 1000, pathLDhat = "", pathPhi = "", format = "fasta", refName = NULL, 
      start = NULL, constant = F, status = T, cores = 1, accept = F, demography = F, out = "")
```

Detailed descriptions of the main functions and all adjacent functions computing the recombination map can be found via e.g.

```R
?LDJump
```

We provide examples with files in *[Example](./Example)*. Previous versions (before version 0.2.1) required a set of lookup-tables of [LDhat](<https://github.com/auton1/LDhat>) which can still be found in *[Lookup Tables](./Lookups)*. However, we recommend to update to the most recent version of **LDJump**. 

A full path to the *Phi* file of [PhiPack](<https://www.maths.otago.ac.nz/~dbryant/software.html>) needs to be provided as follows *pathPhi = "/path/to/Phi"*. In order to use [LDhat](<https://github.com/auton1/LDhat>) to quickly calculate some of the summary statistics please set *pathLDhat = "/path/to/LDhat-master"*.

**LDJump** can also be used under a set of type I error probabilities alpha. Therefore, the parameter *alpha* can also be fed with a vector of values such as:

```R
require(LDJump)
LDJump(seqName, alpha = c(0.1, 0.05, 0.01), segLength = 1000, pathLDhat = "", pathPhi = "", format = "fasta",
       refName = NULL, start = NULL, constant = F, status = T, cores = 1)
```

**LDJump** is designed to estimate recombination rates from segments containing information (by SNPs). Therefore, the program checks all segments (based on the given segment lengths) for the number of SNPS. In case of segments without SNPs, the program will inform the user and ask for input providing the following two options: 
* "n": **LDJump** will be stopped and the user should retry with a larger segment length. 
* "y": **LDJump** will continue and impute the recombination rate of the segment without SNPs with a weighted mean of adjacent segments. 

In order to avoid the user input and accept that **LDJump** imputes recombination rates for these segments, we have added the parameter *accept* (by default *FALSE*), which should then be set to *TRUE*. 

We also included a logical parameter *constant* in **LDJump**, which is *FALSE* by default to estimate **variable recombination rates**. In the case that *constant* is set to *TRUE*, **LDJump** will provide a **constant recombination rate estimator** of the whole sequence. 

A logical parameter *rescale* enables to transform the sequence positions to the unit interval if set to *TRUE*.

An integer parameter *cores* enables to parallelize **LDJump**, where the parameter equals the number of cores on which **LDJump** should run. 

A logical parameter *status*, which is *TRUE* by default, prints the current status of the calculated segment on screen or to the file *LDJump_Status.txt* in case of applying **LDJump** on several cores in parallel. 

A logical parameter *demography*, which is *FALSE* by default, enables to estimate the recombination rates using a generalized additive regression model (GAM) which is based on simulated samples of populations undergoing demographic effects. 

### Update to Version 0.2.2
We have implemented a function *calcRegMod* which computes the regression model for constant estimates by simulating under user defined populations. Therefore, we provide the following example how to obtain the regression  models for populations under a bottleneck followed by rapid growth in population sizes. 

```R
require(LDJump)
scenario =  " -eG 0.0 0 -eG 0.42 -100 -eG 0.5 100 "
simulatedData = calcRegMod(nsim=100,pathToScrm=pathToScrm,scenario=scenario,pathToMs2dna=pathToMs2dna, status = T, pathLDhat = "/path/To/LDhat", pathPhi = "/path/To/Phi")
regMod = simulatedData[[1]]
result = LDJump(seqName = "path/To/seqName", pathLDhat = "/path/To/LDhat", pathPhi = "/path/To/Phi", segLength = 1000, alpha = 0.05, status = T, demography = F, regMod = regMod, status = T, accept = T, cores = 1)
```

Notice that any demographic scenario which is in the range of [scrm](<https://github.com/scrm/scrm/wiki/Command-Line-Options>) can be provided with the simulation function. **However, we want to stress that the resulting model is not checked for its underlying assumptions of the residuals. Nevertheless, this can and should be done manually, given that the model is in hand after the simulations are performed.** Moreover, one can also adapt the sample sizes of the populations (n), the sequence lengths of the simulated populations (len) and the recombination rates. The default values are listed as follows. 
```R
n = c(10, 16, 20)
len = c(500, 1000, 2000, 3000, 5000)
```

For each combination of sample sizes and sequence lengths 100 recombination rates are simulated (uniformly distributed per interval) for the 5 following intervals: [0,0.01], [0.01,0.02], [0.02, 0.05], [0.05, 0.1], and [0.1, 0.2]. As noted above, we added the option to provide a vector of recombination rates under which the recombination map should be simulated. Then the mentioned setup using intervals is ignored. 

Although we did not explore **LDJump** under other scenarios such as population structure, we note that principially (given its implementation in [scrm](<https://github.com/scrm/scrm/wiki/Command-Line-Options>) it is possible to simulate under any scenario using our *calcRegMod* function. **However, one has to be careful that we did not address other scenarios than the demographic example as provided with the R-package. Hence, one has to be careful with LDJump and self-simulated regression models.**

We added an optional parameter *out* which is a prefix for all outfiles created within **LDJump**. By default it is an empty string, but when changed to any user defined-string, one can run **LDJump** in parallel on several data sets from the same working directory without interfering output files. 

### Recommendations
We recommend to run **LDJump** from the same path where the sample file is located in order that all created temp files will be deleted after completion. 

# Reference 
Hermann, P., Heissl, A., Tiemann‐Boege, I., and Futschik, A. (2019), LDJump: Estimating Variable Recombination Rates from Population Genetic Data. Mol Ecol Resour. [doi:10.1111/1755-0998.12994](<https://onlinelibrary.wiley.com/doi/abs/10.1111/1755-0998.12994?af=R>). 


History.md
## Source History

The most recent version of LDJump is 0.2.2 For installation of **LDJump** please see [README.md](README.md) and can be downloaded in this folder. 

### Update Version 0.1.1.
The functions *LDJump* and *getSmuce* are extended with a parameter *constant*. 
By default this is set to *FALSE* such that variable recombination rates are estimated. 
If set to *TRUE*, then a constant recombination rate is estimated for the whole sequence. 

### Update Version 0.1.2
The *smuceR* function to apply SMUCE is exchanged with the function *stepFit* due to the update of the stepR package and the notification that the first function might be removed in future updates. 
Moreover, a note due to package imports and dependencies has been fixed. 

### Update Version 0.1.3
We have changed several function calls due to package imports and dependencies in the previous version. 
We also updated the main function in order to be able to efficiently apply *LDJump* under a vector of type I error probabilities. Moreover, the output has more detailed descriptions. 
We also added an optional parameter to rescale the sequence length to the unit interval. 

### Update Version 0.1.4
Fixed minor bugs and paths to LDhat. Also an optional parameter is added which prints the current status of the calculated segment on screen. Example files are also updated. 

### Update Version 0.1.5
Added the option to impute recombination rates of segments having less than 2 SNPs. User input is required to continue the computation with **LDJump**. 

### Update Version 0.1.6
Fixed minor bug and added option to *fgt_rrate_dpr.R*. 

### Update Version 0.1.7
**LDJump** can now be applied in parallel on several cores. Therefore, an additional parameter is added to the command which reflects the integer number of cores on which the recombination map should be estimated. 

### Update Version 0.1.8
Thanks to user request we have added the option to ignore the user input in case of segments lacking SNPs. Moreover, slight adaptions to commands using the working directory were made. 

### Update Version 0.1.9
We have also simulated under demographic effects and provide a regression model for this setup. This is implemented via an logical parameter *demography*, which will use this regression model if the recombination rates should be estimated under demographic effects. 

### Update Version 0.1.10
Fix issues with paths. *LDJump* should be applied with the full path of the file containing the sequences  (\code{seqName}). Moreover, the function *impute* was rewritten. 

### Update Version 0.2.1
We have implemented the function *calcRegMod* in order to estimate the regression function for constant estimates under simulated populations with a user input demographic scenario. This function requires the two software packages [scrm](<https://github.com/scrm/scrm>) and [ms2dna](<http://guanine.evolbio.mpg.de/bioBox/ms2dna_1.16.tgz>) to be installed. Therefore, we also included an optional parameter in the main function in oder to apply **LDJump** under the newly trained regression model. 

We added the parameter *out* which enables to specify a string added to the output-files. Hence, it is not required anymore to run *LDJump* from different directories when applying it on several data sets in parallel. 

### Update Version 0.2.2
We updated the regression models for constant recombination rates in **LDJump** and therefore, also updated all other models such as the bias correction model or the demography model. These models are based on summary statistics computed from among others the software package [PhiPack](<https://www.maths.otago.ac.nz/~dbryant/software/>). The computing time of **LDJump** is reduced drastically. Additionally, one can still use [LDhat](<https://github.com/auton1/LDhat>) in order to reduce the runtime of **LDJump**.

### Update Version 0.3.1
In this update, we added an extension *vcf_to_fasta* which allows **LDJump** to receive and work with VCF-files. Because of this adjustment, major changes and debugging have been done in other functions of **LDJump**, such as *LDJump*, *summary_statistics*, *vcfR_to_fasta*. Additionally, a new function *vcf_statistics* has been created, which is the main working unit when using VCF-files. The new update also allows the user to parallelize and therefore shorten the running time when working with VCF-files. 



R/calcRegMod.R
calcRegMod = function(n = c(10,16,20), len = c(500,1000,2000,3000,5000), thth = 0.01, nsim = 100, fr = c(), pathToScrm, scenario, pathToMs2dna, status = T, pathLDhat, pathPhi) {
data.all = c()
total = nsim*5*length(n)*length(len)
for(nn in n) {
 for(ll in len) {
  if(length(fr) == 0) {fr = round(c(runif(nsim, 0,0.01),runif(nsim,0.01,0.02),runif(nsim,0.02,0.05),runif(nsim,0.05,0.1),runif(nsim,0.1,0.2)),4)}
  help = c()
  for(j in 1:length(fr)){
   frfr = fr[j];  add = frfr*100000
   frS = frfr*(ll-1)
   thS = thth*(ll-1)
   seqName = paste0(getwd(),"/SimulatedPopulationN",nn,"Len",as.integer(ll),"th",thth,"rr",add,"nsim",j,".fa")
   system(paste0(pathToScrm,"/scrm ",nn," 1 -r ",frS," ",as.integer(ll)," -t ",thS,scenario,"-T | ", pathToMs2dna, "/ms2dna >", seqName))
   while(length(seg.sites(read.dna(seqName, format = "fasta"))) <= 1) {
     system(paste0(pathToScrm,"/scrm ",nn," 1 -r ",frS," ",as.integer(ll)," -t ",thS,scenario,"-T | ", pathToMs2dna, "/ms2dna >", seqName))
   } # end while; loop in order to avoid populations with too little snps
   help = rbind(help, c(unlist(LDJump(seqName, pathLDhat = pathLDhat, pathPhi = pathPhi, format = "fasta", constant = T, status = F, accept = T, regMod = "")[[2]])))
   if(status) {print(paste0(round((max(nrow(data.all)+j, j))/total, 4)*100, "% of in total ", total, " Simulations performed"))}
  } # end fr
 data.all = rbind(data.all, cbind(help, fr))
 }  # end ll
} # end n
# colnames(data.all)[10] = "tajd"
colNam = colnames(data.all)
data.all = data.frame(matrix(unlist(data.all), ncol = 10, byrow = F))
colnames(data.all) = colNam
gamma = 0.5; eps = 0
data.all$fr = ((data.all$fr+eps)^gamma - 1)/gamma
regMod = gam(fr~s(haps)+s(vapw)+s(apwd)+s(hahe)+s(wath)+s(MaxChi)+s(NSS)+s(tajd),data=data.all,family = gaussian(),select = T)
return(list(regMod = regMod, data = data.all))
}
